src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java

```diff
@@ -129,51 +128,71 @@ public synchronized void launch(SlaveComputer computer, TaskListener listener) {
+                                            cloudName,
+                                            namespace,
+                                            pod.getMetadata().getName(),
+                                            e.getMessage());
+                        } else {
+                            node.getRunListener()
+                                    .getLogger()
+                                    .printf(
+                                            "ERROR: Unable to create pod %s %s/%s.%n%s%n",
+                                            cloudName,
+                                            namespace,
+                                            pod.getMetadata().getName(),
+                                            e.getMessage());
+                            PodUtils.cancelQueueItemFor(pod, e.getMessage());
+                        }
+                    } else if (500 <= httpCode && httpCode < 600) { // 5xx
+                        LOGGER.log(FINE, "Kubernetes returned HTTP code {0} {1}. Retrying...", new Object[] {
+                            e.getCode(), e.getStatus()
+                        });
                     } else {
-                        node.getRunListener()
-                                .getLogger()
-                                .printf(
-                                        "ERROR: Unable to create pod %s %s/%s.%n%s%n",
-                                        cloudName, namespace, pod.getMetadata().getName(), e.getMessage());
-                        PodUtils.cancelQueueItemFor(pod, e.getMessage());
+                        LOGGER.log(WARNING, "Kubernetes returned unhandled HTTP code {0} {1}", new Object[] {
+                            e.getCode(), e.getStatus()
+                        });
                     }
-                } else if (500 <= httpCode && httpCode < 600) { // 5xx
-                    LOGGER.log(FINE, "Kubernetes returned HTTP code {0} {1}. Retrying...", new Object[] {
-                        e.getCode(), e.getStatus()
-                    });
-                } else {
-                    LOGGER.log(WARNING, "Kubernetes returned unhandled HTTP code {0} {1}", new Object[] {
-                        e.getCode(), e.getStatus()
-                    });
+                    throw e;
                 }
-                throw e;
+                LOGGER.log(INFO, () -> "Created Pod: " + cloudName + " " + namespace + "/" + podName);
+                listener.getLogger().printf("Created Pod: %s %s/%s%n", cloudName, namespace, podName);
+                Metrics.metricRegistry().counter(MetricNames.PODS_CREATED).inc();
+
+                node.getRunListener().getLogger().printf("Created Pod: %s %s/%s%n", cloudName, namespace, podName);
+            } else {
+                LOGGER.log(INFO, () -> "Pod already exists: " + cloudName + " " + namespace + "/" + podName);
+                listener.getLogger().printf("Pod already exists: %s %s/%s%n", cloudName, namespace, podName);
```

Comment:
So how does this actually differ from https://github.com/jenkinsci/kubernetes-plugin/blob/3b78a0ba311af4c091b94c070b9a8c8b6ab6768b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L106-L110 and do we even still need the `launching` field if we have this logic?

Reply:
That is another face of the same behaviour, just under a slightly different timing (a lot narrower than the one being fixed in this PR).

This PR catches the case there the controller is restarted between the [pod creating call](https://github.com/jenkinsci/kubernetes-plugin/blob/3b78a0ba311af4c091b94c070b9a8c8b6ab6768b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L138) and the end of the [waiting for loop](https://github.com/amuniz/kubernetes-plugin/blob/3b78a0ba311af4c091b94c070b9a8c8b6ab6768b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L279). While the existing `launched` check would only catch a restart right after the [node save](https://github.com/amuniz/kubernetes-plugin/blob/3b78a0ba311af4c091b94c070b9a8c8b6ab6768b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L293) and the end of the method, which is quite unlikely to happen, but possible I guess.

Reply:
>  do we even still need the launching field if we have this logic?

Not sure. https://github.com/amuniz/kubernetes-plugin/blob/f10083a42e708a6bff990c6ccb306ce618b8bccd/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesComputer.java#L201 is never used from this plugin sources, but it's public and it could be used somewhere else, so I would stay on the safe side and keep it.

Reply:
> is never used from this plugin sources, but it's public and it could be used somewhere else

Well, if it is not used it can be deleted. (I usually just do a GH org-wide search.)

Reply:
Then there is stuff like https://github.com/amuniz/kubernetes-plugin/blob/3b78a0ba311af4c091b94c070b9a8c8b6ab6768b/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java#L199-L200 which it would not make sense to do twice, but there is no current defense against race conditions. It seems there are deeper bugs waiting here.